File: src/ua/edu/odeku/ceem/mapRadar/db/models/GeoName.java
package ua.edu.odeku.ceem.mapRadar.db.models;

import org.hibernate.annotations.Index;
import ua.edu.odeku.ceem.mapRadar.db.models.geoNameEnum.*;
import ua.edu.odeku.ceem.mapRadar.exceptions.db.models.GeoNameException;
import ua.edu.odeku.ceem.mapRadar.resource.ResourceString;
import ua.edu.odeku.ceem.mapRadar.settings.PropertyProgram;
import ua.edu.odeku.ceem.mapRadar.utils.translate.TranslateManager;

import javax.persistence.*;
import java.util.Locale;

/**
 * User: Aleo skype: aleo72
 * Date: 09.11.13
 * Time: 20:29
 */
@Entity
@Table(name = "GEO_NAME", uniqueConstraints =
    @UniqueConstraint(columnNames = {"NAME", "LAT", "LON", "FEATURE_CLASS"}))
public class GeoName {

    public GeoName() {
    }

    public GeoName(int sourceId,
                   String name, String asciiname,
                   double lat, double lon,
                   char featureClass, String featureCode) {
        this.sourceId = sourceId;
        this.name = name;
        this.asciiname = asciiname;
        this.lat = lat;
        this.lon = lon;
        this.featureClass = featureClass;
        this.featureCode = featureCode;
        if(PropertyProgram.isTranslateGeoName()){
            this.translateName =
                    TranslateManager.getTranslatable(this.countryCode).translate(this.asciiname);
        } else {
            this.translateName = asciiname;
        }
    }

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    /**
     * integer id of record in geonames database
     * Данное поле необходимо для быстрой проверки на нахождения данного названия в таблице.
     */
    @Column(name = "SOURCE_ID", nullable = false, unique = true)
    @Index(name = "SOURCE_ID_INDEX")
    private int sourceId;

    /**
     * name of geographical point (utf8) varchar(200)
     */
    @Column(name = "NAME", nullable = false, length = 200)
    private String name;

    /**
     * name of geographical point in plain ascii characters, varchar(200)
     */
    @Column(name = "ASCIINAME", nullable = false, length = 200)
    private String asciiname;

    /**
     * alternatenames, comma separated varchar(5000)
     */
    @Column(name = "ALTERNATENAMES", nullable = true, length = 5000)
    private String alternatenames;

    /**
     * latitude in decimal degrees (wgs84)
     */
    @Column(name = "LAT", nullable = false)
    private double lat;

    /**
     * longitude in decimal degrees (wgs84)
     */
    @Column(name = "LON", nullable = false)
    private double lon;

    /**
     * see http://www.geonames.org/export/codes.html, char(1)
     * <p/>
     * A country, state, region,...
     * H stream, lake, ...
     * L parks,area, ...
     * P city, village,...
     * R road, railroad
     * S spot, building, farm
     * T mountain,hill,rock,...
     * U undersea
     * V forest,heath,...
     */
    @Column(name = "FEATURE_CLASS", nullable = false)
    private char featureClass;

    /**
     * see http://www.geonames.org/export/codes.html, varchar(10)
     */
    @Column(name = "FEATURE_CODE", nullable = true)
    private String featureCode;

    /**
     * ISO-3166 2-letter country code, 2 characters
     */
    @Column(name = "COUNTRY_CODE", nullable = false)
    private String countryCode;

    @Column(name = "ELEVATION", nullable = true)
    private Integer elevation;

    @Column(name = "TRANSLATE", nullable = true)
    private String translateName;

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getAsciiname() {
        return asciiname;
    }

    public void setAsciiname(String asciiname) {
        this.asciiname = asciiname;
    }

    public String getAlternatenames() {
        return alternatenames;
    }

    public String[] getAlternatenamesArray() {
        if (alternatenames != null) {
            return alternatenames.split(",");
        }
        return new String[0];
    }

    public void setAlternatenames(String alternatenames) {
        this.alternatenames = alternatenames;
    }

    public void setAlternatenames(String[] alternatenames) {
        if (alternatenames != null) {
            StringBuilder sb = new StringBuilder(alternatenames.length * 20);
            String split = "";
            for (String name : alternatenames) {
                sb.append(split);
                sb.append(name);
                if (split.length() == 0) {
                    split = ",";
                }
            }
            this.alternatenames = sb.toString().trim();
        } else {
            this.alternatenames = null;
        }
    }

    public double getLat() {
        return lat;
    }

    public void setLat(double lat) {
        this.lat = lat;
    }

    public double getLon() {
        return lon;
    }

    public void setLon(double lon) {
        this.lon = lon;
    }

    public char getFeatureClass() {
        return featureClass;
    }

    public void setFeatureClass(char featureClass) throws GeoNameException {
        for(FeatureClass feature : FeatureClass.values()){
            if (feature.toString().equals(String.valueOf(featureClass))) {
                this.featureClass = featureClass;
                return;
            }
        }
        throw new GeoNameException("Feature Class not be used '"+featureClass+"'");
    }

    public String getFeatureCode() {
        return featureCode;
    }

    public void setFeatureCode(String featureCode) throws GeoNameException {
        if (featureCode == null || featureCode.trim().isEmpty())
            return;

        for(FeatureClass feature : FeatureClass.values()){
            if (feature.toString().equals(String.valueOf(featureClass))) {
                switch (feature) {
                    case A:
                        for(FeatureCode_A code : FeatureCode_A.values()){
                            if (code.toString().equals(String.valueOf(featureCode))) {
                                this.featureCode = featureCode;
                                return;
                            }
                        }
                        break;
                    case H:
                        for(FeatureCode_H code : FeatureCode_H.values()){
                            if (code.toString().equals(String.valueOf(featureCode))) {
                                this.featureCode = featureCode;
                                return;
                            }
                        }
                        break;
                    case L:
                        for(FeatureCode_L code : FeatureCode_L.values()){
                            if (code.toString().equals(String.valueOf(featureCode))) {
                                this.featureCode = featureCode;
                                return;
                            }
                        }
                        break;
                    case P:
                        for(FeatureCode_P code : FeatureCode_P.values()){
                            if (code.toString().equals(String.valueOf(featureCode))) {
                                this.featureCode = featureCode;
                                return;
                            }
                        }
                        break;
                    case R:
                        for(FeatureCode_R code : FeatureCode_R.values()){
                            if (code.toString().equals(String.valueOf(featureCode))) {
                                this.featureCode = featureCode;
                                return;
                            }
                        }
                        break;
                    case S:
                        for(FeatureCode_S code : FeatureCode_S.values()){
                            if (code.toString().equals(String.valueOf(featureCode))) {
                                this.featureCode = featureCode;
                                return;
                            }
                        }
                        break;
                    case T:
                        for(FeatureCode_T code : FeatureCode_T.values()){
                            if (code.toString().equals(String.valueOf(featureCode))) {
                                this.featureCode = featureCode;
                                return;
                            }
                        }
                        break;
                    case U:
                        for(FeatureCode_U code : FeatureCode_U.values()){
                            if (code.toString().equals(String.valueOf(featureCode))) {
                                this.featureCode = featureCode;
                                return;
                            }
                        }
                        break;
                    case V:
                        for(FeatureCode_V code : FeatureCode_V.values()){
                            if (code.toString().equals(String.valueOf(featureCode))) {
                                this.featureCode = featureCode;
                                return;
                            }
                        }
                        break;
                }
                break;
            }
        }
        throw new GeoNameException("not valid feature code: " + featureCode);
    }

    public String getCountryCode() {
        return countryCode;
    }

    public void setCountryCode(String countryCode) throws GeoNameException {
        for (String country : Locale.getISOCountries()){
            if(country.equalsIgnoreCase(countryCode)){
                this.countryCode = countryCode.toUpperCase();
                return;
            }
        }
        throw new GeoNameException("not valid countyCode: "+countryCode);
    }

    public Integer getElevation() {
        return elevation;
    }

    public void setElevation(Integer elevation) {
        this.elevation = elevation;
    }

    public int getSourceId() {
        return sourceId;
    }

    public void setSourceId(int sourceId) {
        this.sourceId = sourceId;
    }

    public String getTranslateName() {
        return translateName;
    }

    public void setTranslateName(String translateName) {
        this.translateName = translateName;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(sourceId);
        sb.append(", ");
        sb.append(name);
        sb.append(", ");
        sb.append(asciiname);
        sb.append(", ");
        sb.append(alternatenames);
        sb.append(", ");
        sb.append(lat);
        sb.append(", ");
        sb.append(lon);
        sb.append(", ");
        sb.append(featureClass);
        sb.append(", ");
        sb.append(featureCode);
        sb.append(", ");
        sb.append(countryCode);
        sb.append(", ");
        sb.append(elevation);
        return sb.toString();
    }

    /**
     * Метод преобразует строку в объект.
     * При этом парсит его по указаному стандарту geoName
     * <p/>
     * The main 'geoname' table has the following fields :
     * ---------------------------------------------------
     * geonameid         : integer id of record in geonames database
     * name              : name of geographical point (utf8) varchar(200)
     * asciiname         : name of geographical point in plain ascii characters, varchar(200)
     * alternatenames    : alternatenames, comma separated varchar(5000)
     * latitude          : latitude in decimal degrees (wgs84)
     * longitude         : longitude in decimal degrees (wgs84)
     * feature class     : see http://www.geonames.org/export/codes.html, char(1)
     * feature code      : see http://www.geonames.org/export/codes.html, varchar(10)
     * country code      : ISO-3166 2-letter country code, 2 characters
     * cc2               : alternate country codes, comma separated, ISO-3166 2-letter country code, 60 characters
     * admin1 code       : fipscode (subject to change to iso code), see exceptions below, see file admin1Codes.txt for display names of this code; varchar(20)
     * admin2 code       : code for the second administrative division, a county in the US, see file admin2Codes.txt; varchar(80)
     * admin3 code       : code for third level administrative division, varchar(20)
     * admin4 code       : code for fourth level administrative division, varchar(20)
     * population        : bigint (8 byte int)
     * elevation         : in meters, integer
     * dem               : digital elevation model, srtm3 or gtopo30, average elevation of 3''x3'' (ca 90mx90m) or 30''x30'' (ca 900mx900m) area in meters, integer. srtm processed by cgiar/ciat.
     * timezone          : the timezone id (see file timeZone.txt) varchar(40)
     * modification date : date of last modification in yyyy-MM-dd format
     *
     * @param line строка для парсинга
     * @return GeoName объект
     */
    public static GeoName createGeoName(String line) throws GeoNameException {
        GeoName geoName = new GeoName();

        String[] columns = line.trim().split("\t");

        try{
            geoName.setSourceId(Integer.parseInt(columns[0]));
            geoName.setName(columns[1]);
            geoName.setAsciiname(columns[2]);
            geoName.setAlternatenames(columns[3]);
            geoName.setLat(Double.parseDouble(columns[4]));
            geoName.setLon(Double.parseDouble(columns[5]));
            geoName.setFeatureClass(columns[6].charAt(0));
            geoName.setFeatureCode(columns[7]);
            geoName.setCountryCode(columns[8]);
            String eleva = columns[15];
            if(eleva != null && !eleva.trim().isEmpty())
                geoName.setElevation(Integer.parseInt(eleva));

            if(PropertyProgram.isTranslateGeoName()){
                geoName.translateName =
                        TranslateManager.getTranslatable(geoName.countryCode).translate(geoName.asciiname);
            } else {
                geoName.translateName = geoName.asciiname;
            }

        } catch (Exception e){
            throw new GeoNameException(ResourceString.get("exception_GeoName")+"\n"+ line + "\n" + e.getMessage());
        }

        return geoName;
    }


}

File: src/ua/edu/odeku/ceem/mapRadar/db/model/GeoName.scala
/*
 * Odessa State environmental University
 * Copyright (C) 2014
 */

package ua.edu.odeku.ceem.mapRadar.db.model

import gov.nasa.worldwind.geom.LatLon
import ua.edu.odeku.ceem.mapRadar.db.{CeemTableObject, DB}
import ua.edu.odeku.ceem.mapRadar.exceptions.db.models.GeoNameException
import ua.edu.odeku.ceem.mapRadar.resource.ResourceString
import ua.edu.odeku.ceem.mapRadar.settings.PropertyProgram

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer
import scala.slick.driver.H2Driver
import scala.slick.driver.H2Driver.simple._
import scala.slick.jdbc.meta.MTable
import scala.slick.jdbc.{StaticQuery => Q}

/**
 * User: Aleo Bakalov
 * Date: 04.07.2014
 * Time: 12:58
 */

case class GeoName(id: Long,
                   name: String,
                   ascii: String,
                   alternateNames: String,
                   lat: Double,
                   lon: Double,
                   featureClass: String,
                   featureCode: String,
                   countryCode: String,
                   translateName: Option[String] = None) {

	val fields = Array(id, name, ascii, translateName, alternateNames, countryCode, featureClass, featureCode, lat, lon)
}

class GeoNames(tag: Tag) extends Table[GeoName](tag, "GEO_NAMES") {
	/**
	 * integer id of record in geonames database
	 * Данное поле необходимо для быстрой проверки на нахождения данного названия в таблице.
	 */
	def id = column[Long]("ID", O.PrimaryKey)

	/**
	 * name of geographical point (utf8) varchar(200)
	 */
	def name = column[String]("NAME", O.NotNull)

	/**
	 * name of geographical point in plain ascii characters, varchar(200)
	 */
	def ascii = column[String]("ASCII_NAME", O.NotNull)

	/**
	 * alternatenames, comma separated varchar(5000)
	 */
	def alternateNames = column[String]("ALTERNATE_NAMES", O.NotNull)

	/**
	 * latitude in decimal degrees (wgs84)
	 */
	def lat = column[Double]("LAT", O.NotNull)

	/**
	 * longitude in decimal degrees (wgs84)
	 */
	def lon = column[Double]("LON", O.NotNull)

	/**
	 * see http://www.geonames.org/export/codes.html, char(1)
	 * <p/>
	 * A country, state, region,...
	 * H stream, lake, ...
	 * L parks,area, ...
	 * P city, village,...
	 * R road, railroad
	 * S spot, building, farm
	 * T mountain,hill,rock,...
	 * U undersea
	 * V forest,heath,...
	 */
	def featureClass = column[String]("FEATURE_CLASS", O.NotNull)

	/**
	 * see http://www.geonames.org/export/codes.html, varchar(10)
	 */
	def featureCode = column[String]("FEATURE_CODE", O.Nullable)

	/**
	 * ISO-3166 2-letter country code, 2 characters
	 */
	def countryCode = column[String]("COUNTRY_CODE", O.NotNull)

	def translateName = column[Option[String]]("TRANSLATE", O.Nullable)

	def alternateNamesOfArray = if (alternateNames == null) Array() else alternateNames.toString().split(",")

	override def * = (id, name, ascii, alternateNames, lat, lon, featureClass, featureCode, countryCode, translateName) <>(GeoName.tupled, GeoName.unapply)
}

object GeoNames extends CeemTableObject {


	val countFieldsInGeoName = 10

	val objects = TableQuery[GeoNames]

	val table = GeoNames.objects.baseTableRow

	def tableName: String = this.objects.baseTableRow.tableName

	def geoNameTable = tableName
	def nameColumn = table.name.toString()
	def translateNameColumn = table.translateName.toString()
	def latColumn = table.lat.toString()
	def lonColumn = table.lon.toString()
	def featureClassColumn = table.featureClass.toString()
	def countryColumn = table.countryCode.toString()
	def featureCodeColumn = table.featureCode.toString()

	override def createIfNotExists(existsTables: List[MTable], db: H2Driver.backend.DatabaseDef = DB.database) {
		if (!existsTables.exists(_.name.name == tableName)) {
			db withSession { implicit session =>
				if (PropertyProgram.DEBUG) println(objects.ddl.createStatements.mkString("\n"))
				objects.ddl.create
			}
		}
	}

	def += (geoName: GeoName) = {
		DB.database withSession { implicit session =>
			TableQuery[GeoNames] += geoName
		}
	}

	def ++=(values: Iterable[GeoName]) = {
		DB.database withSession { implicit session =>
			TableQuery[GeoNames] ++= values
		}
	}

	def -= (geoName: GeoName): Int = {
		DB.database withSession { implicit session =>
			if(PropertyProgram.DEBUG) println(TableQuery[GeoNames].filter(_.id === geoName.id).deleteStatement)
			TableQuery[GeoNames].filter(_.id === geoName.id).delete
		}
	}

	def ! (geoName: GeoName): Int = {
		DB.database withSession { implicit session =>
			if(PropertyProgram.DEBUG) println(TableQuery[GeoNames].filter(_.id === geoName.id).updateStatement)
			TableQuery[GeoNames].filter(_.id === geoName.id).update(geoName)
		}
	}

	def createGeoName(line: String) = {
		val c = line.split("\t")

		try {
			GeoName(
			id = c(0).toInt,
			name = c(1),
			ascii = c(2),
			alternateNames = c(3),
			lat = c(4).toDouble,
			lon = c(5).toDouble,
			featureClass = c(6),
			featureCode = c(7),
			countryCode = c(8)
			)
		} catch {

			case e: Exception =>
				val message = ResourceString.get("exception_GeoName")
				val exceptionMessage = e.getMessage
				throw new GeoNameException(s"$message\n$line\n$exceptionMessage")
		}
	}

	def get(id: Long): GeoName = {
		DB.database withSession {implicit session =>
			if(PropertyProgram.DEBUG) println(TableQuery[GeoNames].filter(_.id === id).selectStatement)
		  TableQuery[GeoNames].filter(_.id === id).first
		}
	}

	def list(subName: String = null, country: String = null, featureClass: String = null, featureCode: String = null) = {

		var query: Query[GeoNames, GeoNames#TableElementType, Seq] = null

		query = if (country != null && !country.isEmpty) objects.filter(_.countryCode === country) else query

		query = if (featureClass != null && !featureClass.isEmpty) {
			query = if (query == null) objects.filter(_.featureClass === featureClass) else query.filter(_.featureClass === featureClass)
			query = if (featureCode != null && !featureCode.isEmpty) query.filter(_.featureCode === featureCode) else query
			query
		} else query

		query = if (subName != null && !subName.trim.isEmpty) {
			val nameLike = s"%$subName%"
			query.filter(_.name like nameLike) union query.filter(_.translateName like nameLike) union query.filter(_.ascii like nameLike) union query.filter(_.alternateNames like nameLike)
		} else query

		query = if (query == null) GeoNames.objects.filter(_.id >= 0L) else query

		query = query.sortBy(_.alternateNames).sortBy(_.ascii).sortBy(_.name).sortBy(_.translateName)

		DB.database withSession {
			implicit session =>

				if (PropertyProgram.DEBUG) println(query.selectStatement)

				query.list.toArray
		}
	}

	/**
	 * Возращает список feature codes
	 * @param featureClass класс фич, из которых необходимо выбрать коды фич
	 * @return список код фич
	 */
	def featureCodes(featureClass: String): List[String] = {
		val selectedColumn = featureCodeColumn
		DB.database withSession { implicit session =>
			val sql = Q.queryNA[String](
				s"""
					SELECT
						$selectedColumn
		      FROM
						$geoNameTable
					WHERE
		        $selectedColumn IS NOT NULL
						AND $featureClassColumn = '$featureClass'
					GROUP BY
		        $selectedColumn
					ORDER BY
						$selectedColumn
				 """)
			sql.list
		}
	}

	def featureClass: List[String] = {
		val selectedColumn = featureClassColumn
		DB.database withSession { implicit session =>
			val sql = Q.queryNA[String](
				s"""
					SELECT
						$selectedColumn
		      FROM
						$geoNameTable
					WHERE
		        $selectedColumn IS NOT NULL
					GROUP BY
		        $selectedColumn
					ORDER BY
						$selectedColumn
				 """)
			sql.list
		}
	}

	def coutres: List[String] = {
		val selectedColumn = countryColumn
		DB.database withSession { implicit session =>
			val sql = Q.queryNA[String](
				s"""
					SELECT
						$selectedColumn
		      FROM
						$geoNameTable
					WHERE
		        $selectedColumn IS NOT NULL
					GROUP BY
		        $selectedColumn
					ORDER BY
						$selectedColumn
				 """)
			sql.list
		}
	}

	object SavingCache {
		private val list = new ArrayBuffer[GeoName]()

		val LIMIT = 500

		def += (geo: GeoName): Unit = {
			list += geo

			if(list.size >= LIMIT){
				saveCache()
			}
		}

		def saveCache() {
			GeoNames ++= list
			list.clear()
		}

		def flush(): Unit = saveCache()
	}
}

case class GeoNamesWithNameAndCoordinates(name: String, latlon: LatLon) {

	override def toString = name
}

object GeoNamesWithNameAndCoordinates {
	import ua.edu.odeku.ceem.mapRadar.db.model.GeoNames._

	def getSettlements(text: String): List[GeoNamesWithNameAndCoordinates] = {

		val sql = Q.queryNA[(String, Double, Double)](
			s"""
			  SELECT
			    CASE
			      WHEN $translateNameColumn IS NOT NULL
			        THEN $translateNameColumn
			        ELSE NAME
			    END AS $nameColumn,

			    $latColumn,
			  	$lonColumn
			  FROM
			    $geoNameTable
				WHERE
					$featureClassColumn != 'A'
			  	AND (
						UPPER($nameColumn) LIKE '$text'
							OR
						UPPER($translateNameColumn) LIKE '$text'
			  	)
			  	ORDER BY
			  				$nameColumn;
			""")

		val list = new mutable.MutableList[GeoNamesWithNameAndCoordinates]

		DB.database withSession { implicit session =>
			sql foreach {
				v => list += GeoNamesWithNameAndCoordinates(v._1, LatLon.fromDegrees(v._2, v._3))
			}
		}
		list.toList
	}
}


object Test extends App {
	//	DB.database withSession {
	//		implicit session =>
	//			GeoNames.objects.ddl.drop
	//			GeoNames.objects.ddl.create
	//
	//	}
	for (i <- 1 to 1000) {
		GeoNames += GeoName(i, "test" + i, i + "test2", i + "test3,test4" + i, 7234.90, 7294.94, "F", "FF", "UA", null)
	}

	//	GeoNames += GeoName(1, "test", "test2", "test3,test4", 7234.90, 7294.94, "F", "FF", "UA", null)

	println(GeoNames.list("", "UA", "FF", "FF").mkString(","))
	println(GeoNames.list("st4", "UA", "F", "FF").mkString(","))

	print(GeoNames.list().mkString(","))
}